### PR TITLE
81 styled landing page

### PR DESCRIPTION
**Pull request by jaredhirsch**
@lmorchard @meandavejustice r?

Pull in @johngruen's styles + html from the johngruen/idea-town-scratchpad repo, and also split out the templates into separate files, to keep the JS views concise.

### CODE REVIEW

File: idea_town/frontend/static-src/app/views/experiment-page.js [hunk 1]
```diff
@@ -1,17 +1,10 @@
 import app from 'ampersand-app';
 
 import BaseView from './base-view';
+import template from '../templates/experiment-page.js';
```
Comment (pdehaan) [inline idea_town/frontend/static-src/app/views/experiment-page.js]: Nit: remove the trailing `.js` for consistency.

File: idea_town/frontend/static-src/app/templates/landing-page.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  <div>
+  {{^loggedIn}}
+    <div class="firefox-logo"></div>
+    <h1 class="hero">Introducing Idea Town</h1>
+    <h2 class="sub-hero">We're building the next generation of<br> Firefox features and we want your feedback! <br>Get started with a Firefox Account.</h2>
+    <div class="cta-layout-wrapper">
+      <div class="cta-layout">
+        <a href="/accounts/login/?next=/home"><button class="button large primary">Sign up</button></a>
+        <a href="/accounts/login/?next=/home" class="fxa-alternate">Already have an account? Sign in.</a>
+        <p class="cta-legal">By proceeding, you agree to the Terms of Service and Privacy Notice of Idea Town.</p>
```
Comment (pdehaan) [inline idea_town/frontend/static-src/app/templates/landing-page.js]: Should "Terms of Service" and "Privacy Notice" be linked here (even if just to "#" for now).

File: idea_town/frontend/static-src/app/views/header-view.js [hunk 3]
```diff
@@ -2,11 +2,10 @@ import app from 'ampersand-app';
 import xhr from 'xhr';
 
 import BaseView from './base-view';
+import template from '../templates/header-view.js';
```
Comment (pdehaan) [inline idea_town/frontend/static-src/app/views/header-view.js]: Nit: remove the trailing `.js` for consistency.

File: idea_town/frontend/static-src/styles/modules/_cta.scss [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    .cta-layout {
+      flex: 1 1;
+
+      .button {
+        width: 100%;
+      }
+
+      .fxa-alternate {
+        color: #fff;
```
Comment (pdehaan) [inline idea_town/frontend/static-src/styles/modules/_cta.scss]: `#fff` => `$white`

File: idea_town/frontend/templates/idea_town/frontend/index.html [hunk 5]
```diff
@@ -2,24 +2,19 @@
 <html>
 <head>
     <title>Idea Town</title>
+    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700,300' rel='stylesheet' type='text/css'>
```
Comment (pdehaan) [inline idea_town/frontend/templates/idea_town/frontend/index.html]: Ultra nit: Double quotes for consistency.

File: idea_town/frontend/static-src/app/templates/not-found-page.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+export default `
+  <section class="page" data-hook="not-found-page">
+    <h1>404 wut</h1>
+  </section>`
```
Comment (lmorchard) [inline idea_town/frontend/static-src/app/templates/not-found-page.js]: Extra stray <code>`</code> here causes an "Unterminated template" error for me on build

File: idea_town/frontend/static-src/app/views/landing-page.js [hunk 7]
```diff
@@ -9,25 +10,27 @@ const addonInfo = {
...
+
+    // TODO: setting styles and ids on body seems funny
+    if (app.me.session) {
+      document.body.classList.add('add-on');
+    } else {
+      document.body.classList.add('sign-up');
+    }
+    // save the old ID, if there was one
+    document.body._id = document.body.id;
+    document.body.id = "cta";
```
Comment (pdehaan) [inline idea_town/frontend/static-src/app/views/landing-page.js]: nit: single quotes for the eslint gods.

File: idea_town/frontend/static-src/app/views/not-found-page.js [hunk 8]
```diff
@@ -1,7 +1,6 @@
 import BaseView from './base-view';
+import templtae from '../templates/not-found-page';
```
Comment (pdehaan) [inline idea_town/frontend/static-src/app/views/not-found-page.js]: Typo: `templtae` => `template`

```
Users/pdehaan/dev/tmp/del/6a68_ideatown/idea_town/frontend/static-src/app/views/not-found-page.js
  2:8   error  "templtae" is defined but never used  no-unused-vars
  5:14  error  "template" is not defined
```

File: idea_town/frontend/static-src/app/templates/not-found-page.js (hunk 6 above)
Comment (pdehaan) [inline idea_town/frontend/static-src/app/templates/not-found-page.js]: Filed #87 to track whether we can get our Travis container to also run some Node+Gulp tasks.
I haven't played too much with multi-language Travis projects yet. /cc @nchapman (whom I think did this with "Find My Device" w/ Go+Node)

Comment (nchapman) [inline idea_town/frontend/static-src/app/templates/not-found-page.js]: My recollection is that you can only specifically target a single language, but that other (popular?) languages are available for use (just can't force a specific version).

File: idea_town/frontend/static-src/app/templates/landing-page.js (hunk 2 above)
Comment (jaredhirsch) [inline idea_town/frontend/static-src/app/templates/landing-page.js]: Sure!

**Comment by jaredhirsch** [comment]
Thanks for the incredibly thorough syntactic analysis, @pdehaan ^_^

Pretty sure I've addressed all the nits, micro-nits, and nano-nits. Waiting for docker to build locally to be sure nothing else is busted.

**Comment by pdehaan** [comment]
OK, I got this running again locally...
Everything looks great! Awesome job @johngruen and @6a68!

I think this is safe to merge, and we can address any future work in a separate PR. Here are some casual things that I noticed (plus some screenshots for the curious travelers to this repo):
1. Currently it looks like "Sign up" and "Sign in" both do the same thing.
2. "Terms of Services" and "Privacy Notice" should be linked on the homepage, but not sure where to direct those yet.
3. Needs some responsive breakpoints as this doesn't re-flow very well on non-desktops (or desktop browsers with smaller browsers). For example, text isn't wrapping and the Tabzilla icon gets lost off the edge of the screen.
4. No Tabzilla tab on the "Install the Add-on" page.
5. The "Install the Add-on" button seems to download an .XPI, which I had to track down in Finder, then gives me the "This add-on could not be installed because it has not been verified." notice.

---

idea_town

idea_town

After installing the add-on, it seems to take me to "/home" route, but it still prompts me to install the add-on and doesn't seem to have any styling:

idea_town

And here's what I was saying about breakpoints and stuff. Note how the text is flush with the left edge, and flows merrily off the right of the screen instead of wrapping cleanly.

idea_town

**Comment by pdehaan** [comment]
Unrelated, but here's some stats on some image assets. Oddly, it looks like there are 2 places where the high-res `@2x` images are smaller than their low res counterparts.
Also odd is that "/idea_town/frontend/static/media/img/town@2x.png" is 211KB, yet it seems like it's still serving me the low-res 57KB version on my MBP, so I may need to peek into the hidpi media queries.

```
$ pwd && ls -lh
/Users/pdehaan/dev/tmp/del/6a68_ideatown/idea_town/frontend/static/media/img
total 16
-rw-r--r--  1 pdehaan  staff   1.2K Sep 25 12:28 tabzilla-static-high-res.png
-rw-r--r--  1 pdehaan  staff   2.2K Sep 25 12:28 tabzilla-static.png
```

```
$ pwd && ls -lh
/Users/pdehaan/dev/tmp/del/6a68_ideatown/idea_town/frontend/static/images
total 840
-rw-r--r--  1 pdehaan  staff   622B Sep 25 12:29 count-avatar.png
-rw-r--r--  1 pdehaan  staff   1.1K Sep 25 12:29 count-avatar@2x.png
-rw-r--r--  1 pdehaan  staff   1.2K Sep 25 12:29 default-avatar.png
-rw-r--r--  1 pdehaan  staff   2.1K Sep 25 12:29 default-avatar@2x.png
-rw-r--r--  1 pdehaan  staff    28K Sep 25 12:29 firefox-logo.png
-rw-r--r--  1 pdehaan  staff    20K Sep 25 12:29 firefox-logo@2x.png
-rw-r--r--  1 pdehaan  staff   8.6K Sep 25 12:29 grad-background.png
-rw-r--r--  1 pdehaan  staff    23K Sep 25 12:29 grad-background@2x.png
-rw-r--r--  1 pdehaan  staff   4.1K Sep 25 12:29 icon-64.png
-rw-r--r--  1 pdehaan  staff   1.4K Sep 25 12:29 mozilla.png
-rw-r--r--  1 pdehaan  staff   2.4K Sep 25 12:29 mozilla@2x.png
-rw-r--r--  1 pdehaan  staff   6.0K Sep 25 12:29 small-firefox-logo.png
-rw-r--r--  1 pdehaan  staff    17K Sep 25 12:29 small-firefox-logo@2x.png
-rw-r--r--  1 pdehaan  staff    57K Sep 25 12:29 town.png
-rw-r--r--  1 pdehaan  staff   211K Sep 25 12:29 town@2x.png
```

**Comment by pdehaan** [comment]
Can't remember if we discussed this in a separate PR/bug or not yet, but we should also maybe [re]discuss the optional trailing slash behavior.

(mainly because I just wasted 5 minutes trying to remember if it was `/accounts/logout` or `/accounts/logout/` because of a bad auto-complete in my browser history)
- http://ideatown.dev:8000/home is good.
- http://ideatown.dev:8000/home/ gives a a "404 wut" error page:
  idea_town

**Comment by jaredhirsch** [comment]
@pdehaan Thanks for the thorough review!

Stuff you mentioned:

> 1. Currently it looks like "Sign up" and "Sign in" both do the same thing.

Yup, though I'm not sure if django-allauth's fxa support exposes that via a querystring. Filed as #89

> 2 "Terms of Services" and "Privacy Notice" should be linked on the homepage, but not sure where to direct those yet.

Try refreshing your branch; I've linked to the existing Cloud Services ToS/privacy pages, just to have some kind of placeholder.

> 3 Needs some responsive breakpoints as this doesn't re-flow very well on non-desktops (or desktop browsers with smaller browsers). For example, text isn't wrapping and the Tabzilla icon gets lost off the edge of the screen.

Filed as #90

> 4 No Tabzilla tab on the "Install the Add-on" page.

The tabzilla tab on the "install addon" page is there, it's just white-on-white. Filed #91 to fix it.

> 5 The "Install the Add-on" button seems to download an .XPI, which I had to track down in Finder, then gives me the "This add-on could not be installed because it has not been verified." notice.

Yeah, this is because we don't have an AMO link in there yet. Filed as #92.

> After installing the add-on, it seems to take me to "/home" route, but it still prompts me to install the add-on and doesn't seem to have any styling

This is a legit bug. The problem was that I had overridden the base `remove()` method in the `LandingPage` view, but wasn't calling the method on the "superclass". Should be fixed once I push.

> Can't remember if we discussed this in a separate PR/bug or not yet, but we should also maybe [re]discuss the optional trailing slash behavior.

Django assumes trailing slashes, but it's configured not to require them. Ampersand assumes no trailing slashes, but I can add a little something to the Ampersand router so that it tolerates trailing slashes.

I'll push fixes momentarily.

**Comment by jaredhirsch** [comment]
For some reason, the header isn't updating itself properly on login (you only rarely see a 'log out' button above the experiment list on the `/home` page). I don't know what the issue is, but I'm tired of this PR, so I'm filing followup bug #95 to figure it out.

Should (hopefully) be good to land. @lmorchard mind giving the final once-over?

**Comment by pdehaan** [comment]
@6a68: this is awesome!

The splash page looks great in Chrome (we'll still need to add styling for the ToS/PN CTA buttons, but this is a great start).

Not sure why my [signed in] Firefox is misbehaving, but it's currently showing me a blank-ish page with the message:

> ### "TypeError: `this.beforeRender` is not a function"
> 
> &mdash; [app.js:544:5](https://github.com/mozilla/idea-town/blob/2e266451530ec119fc2a11019aedf82dea25153d/idea_town/frontend/static-src/app/views/base-view.js#L20-L24)

---

idea_town

**Comment by meandavejustice** [comment]
Getting this error on nearly every page. 
`TypeError: this.beforeRender is not a function app.js:544:4`

**Comment by jaredhirsch** [comment]
@meandavejustice @pdehaan I'm not seeing those errors right now, though I've seen similar errors in the past. I'd bet that running the steps in my [mediocre shell script](https://github.com/mozilla/idea-town/pull/98) will get you a working build. Ping me on IRC if not, and we can debug in real time.

**Comment by jaredhirsch** [comment]
Also note that, if you go to the admin site (`ideatown.dev:8000/admin/`, username + password are `admin`), and create a couple of experiments, you'll see them rendered on the `/home` page after you login with the addon installed.

**Comment by jaredhirsch** [comment]
Here's the `/home` page from this patch with no experiments:

screenshot 2015-09-28 12 14 48

And with two demo experiments:

screenshot 2015-09-28 12 14 59

No errors in the console :-\

**Comment by jaredhirsch** [comment]
Just for reference, the screenshotted gifs are:

wowow

cheese-2

#blessed #toomanygifs

**Comment by lmorchard** [comment]
I can confirm that all the above works for me. 

The one weird nitpicky thing is I see a flash of content of the "install addon plz" message before the experiments show up. Having picked at that nit, I'm not sure at all how we could fix that. We get a message from an installed addon - or we get nothing. So, it makes sense to show the no-addon case and replace it when the addon is around.

Oh, one possibility though: If we're currently waiting until the API has responded with experiments to switch views, it might be handy to immediately switch to a "loading" view on addon detection and _then_ switch to the experiments listing view

**Comment by lmorchard** [comment]
I do also get the `this.beforeRender` error. But, I'm going to try rebuilding my docker containers to see if that changes things. Running at least `docker-compose build && docker-compose restart` after any `git pull` is a good practice to have. (Which I didn't do until just now)

**Comment by jaredhirsch** [comment]
> I do also get the this.beforeRender error

Pretty weird case of "only works on developer's machine." I'm really curious if rebuilding will fix it.

**Comment by jaredhirsch** [comment]
> The one weird nitpicky thing is I see a flash of content of the "install addon plz" message before the experiments show up

Here's STR, I think I might be seeing the same thing:
- have the addon
- then, login
- on redirect, see a flash of the landing page before the home page is shown

Is that the STR that do it for you? If I just type in `ideatown.dev:8000/home` I don't see the landing page flash in the background.

At any rate, I'd probably like to file followup bugs for elusive optimization-ish bugs like this one, if only because this PR's been sitting for a while. Does that seem like a reasonable course of action?

**Comment by lmorchard** [comment]
Hmm, rebuilding didn't fix the beforeRender thing, and the place it's happening is bizarre. I shoved some console.logs into header-view like so:

``` diff
--- a/idea_town/frontend/static-src/app/views/header-view.js
+++ b/idea_town/frontend/static-src/app/views/header-view.js
@@ -22,7 +22,9 @@ export default BaseView.extend({
     // an active user has an addon and a session
     this.activeUser = !!app.me.user.id && app.me.hasAddon;

+    console.log("HURF", this);
     BaseView.prototype.render.apply(this, arguments);
+    console.log("DURF", this);
   },
```

And I get errors like so:

```
HURF Object { cid: "state1", _events: Object, _values: Object, _definition: Object, parent: undefined, collection: undefined, _keyTree: Object, _cache: Object, _previousAttributes: Object, _changing: true, 4 more… } app.js:726:5
TypeError: this.beforeRender is not a function
```

That's especially bizarre - because in that method in that view, `this` has apparently turned into a copy of the `app.me` state object when I click on the Object in the console message.

I think there's probably an onChange event handler somewhere that's doing something odd.

**Comment by lmorchard** [comment]
Oh! Here you go, this seems to fix it:

``` diff
--- a/idea_town/frontend/static-src/app/views/header-view.js
+++ b/idea_town/frontend/static-src/app/views/header-view.js
@@ -13,7 +13,7 @@ export default BaseView.extend({
   initialize() {
     // addon changes may be broadcast by the idea-town addon outside the
     // regular page load cycle, and those changes alter the header's appearance
-    app.me.on('change:hasAddon', this.render);
+    app.me.on('change:hasAddon', this.render, this);
   },
```

In that event handler, `this` turns into the state object. And so without that 3rd parameter, the view's `render` gets called as bound to the state, just as the error suggested.

**Comment by jaredhirsch** [comment]
@lmorchard Well, this is fun! I haven't had to blindly try to debug errors I can't reproduce since I was submitting gaia builds to try :poop:.

I made the edit you suggested and force-pushed. Maybe try again?

**Comment by lmorchard** [comment]
Yup, that seemed to fix it - and also now the "logged in as" thing at the top appears for me!

**Comment by lmorchard** [comment]
Spent some more time clicking around and browsing through the changes. Looks good to me. Let's get this big bite chewed & swallowed!

**Pull request merged**